frontend/dockerfile/instructions/parse.go

```diff
@@ -66,8 +69,13 @@ func newParseRequestFromNode(node *parser.Node) parseRequest {
 	}
 }
 
-// ParseInstruction converts an AST to a typed instruction (either a command or a build stage beginning when encountering a `FROM` statement)
 func ParseInstruction(node *parser.Node) (v interface{}, err error) {
+	noWarn := func(string, string, [][]byte, *parser.Range) {}
+	return ParseInstructionWithWarnings(node, noWarn)
+}
+
+// ParseInstruction converts an AST to a typed instruction (either a command or a build stage beginning when encountering a `FROM` statement)
+func ParseInstructionWithWarnings(node *parser.Node, warn warnCallback) (v interface{}, err error) {
```

Comment:
Private types in public methods should be avoided.

It may make sense to turn these to methods of `Parser` that can have the warner configured so it doesn't need to be passed all the time.

Reply:
I ended up refactoring this as part of getting the better short messages in, but this is a really good callout for me, as I hadn't even considered that.